### PR TITLE
Optimized level change updator

### PR DESCRIPTION
**Pull request by gavinnwang**

### CODE REVIEW

File: exchange/test/src/unit/matching/get_update.cpp [hunk 1]
```diff
@@ -19,6 +20,9 @@ class UnitGetUpdate : public ::testing::Test {
     using TestTrader = nutc::test_utils::TestTrader;
     std::shared_ptr<nutc::traders::GenericTrader> trader1, trader2, trader3;
 
+    std::shared_ptr<LevelUpdateGenerator> generator_ =
```
[the remark this answers is not in the document]
Comment (gavinnwang) [reply exchange/test/src/unit/matching/get_update.cpp:23]: this is a class member so i dont think we can use auto here

File: exchange/test/src/unit/matching/get_update.cpp [hunk 2]
```diff
@@ -132,21 +143,22 @@ TEST_F(UnitGetUpdate, ManyLevelChanges)
...
-    after.add_order(order8);
-
-    auto updates = get_updates("A", before, after);
+
+    ob.add_order(order1);
+    ob.add_order(order2);
+    ob.add_order(order3);
+    ob.add_order(order4);
+
+    generator_->reset();
```
[the remark this answers is not in the document]
Comment (gavinnwang) [reply exchange/test/src/unit/matching/get_update.cpp:152]: the reset is called every tick so this is basically for mimicking the behavior of a new tick

**Comment by egelja** [comment]
i have some comments, will leave tomorrow.

File: exchange/test/src/unit/matching/get_update.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevenewald) [reply exchange/test/src/unit/matching/get_update.cpp:23]: oh ur right

File: exchange/test/src/unit/matching/get_update.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stevenewald) [reply exchange/test/src/unit/matching/get_update.cpp:153]: ah ic

File: exchange/src/exchange/tickers/engine/level_update_generator.cpp [hunk 3]
```diff
@@ -1,57 +1,35 @@
...
+    return updates;
+}
 
-        updates.emplace_back(ticker, util::Side::buy, price, after_quantity);
+void
+LevelUpdateGenerator::record_level_change(
+    util::Side side, double price, double new_quantity
+)
+{
+    assert(std::round(price * 100) / 100 == price);
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/level_update_generator.cpp:26]: you may hit FPE bugs here

File: exchange/src/exchange/tickers/engine/level_update_generator.hpp [hunk 4]
```diff
@@ -1,15 +1,28 @@
...
 namespace matching {
 using ob_update = messages::orderbook_update;
 
-std::vector<ob_update>
-get_updates(const std::string& ticker, const OrderBook& before, const OrderBook& after);
+class LevelUpdateGenerator {
+public:
+    void record_level_change(util::Side side, double price, double new_quantity);
 
+    std::vector<ob_update> get_updates(const std::string& ticker);
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/level_update_generator.hpp:14]:
```suggestion
    std::vector<ob_update> get_updates(const std::string& ticker) const;
```

File: exchange/src/exchange/tickers/engine/order_container.cpp [hunk 5]
```diff
@@ -43,6 +49,34 @@ OrderBook::remove_order(uint64_t order_id)
...
+    if (bids_.empty() || asks_.empty()) {
+        return 0;
+    }
+    return (bids_.begin()->price + asks_.begin()->price) / 2;
+}
+
+double
+OrderBook::get_level(util::Side side, double price) const
+{
+    const auto& levels = side == util::Side::buy ? bid_levels_ : ask_levels_;
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/order_container.cpp:73]:
```suggestion
    const auto& levels = (side == util::Side::buy) ? bid_levels_ : ask_levels_;
```

File: exchange/src/exchange/tickers/engine/order_container.cpp [hunk 6]
```diff
@@ -43,6 +49,34 @@ OrderBook::remove_order(uint64_t order_id)
...
+    if (levels.find(price) == levels.end()) {
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/order_container.cpp:74]:
```suggestion
    if (!levels.contains(price)) {
```

File: exchange/src/exchange/tickers/engine/order_container.hpp [hunk 7]
```diff
@@ -27,61 +30,18 @@ class OrderBook {
...
-     * @brief Get the price->quantity map for a SIDE
-     */
-    const std::unordered_map<double, double>&
-    get_levels(util::Side side) const
-    {
-        return side == util::Side::buy ? bid_levels_ : ask_levels_;
-    }
+    OrderBook(std::shared_ptr<LevelUpdateGenerator> level_update_generator = nullptr) :
+        level_update_generator_(level_update_generator)
+    {}
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/order_container.hpp:35]: you should rarely use default parameters. Overloading is almost always a better solution

```suggestion
    explicit OrderBook() : level_update_generator_(nullptr) {}

    explicit OrderBook(std::shared_ptr<LevelUpdateGenerator> level_update_generator) :
        level_update_generator_(level_update_generator)
    {}
```

File: exchange/src/exchange/tickers/manager/ticker_manager.cpp [hunk 8]
```diff
@@ -41,6 +41,16 @@ log_match(const matching::stored_match& order)
...
 }
 
+size_t
+EngineManager::match_order(const matching::stored_order& order)
+{
+    auto& ticker = get_engine(order.ticker);
+    std::vector<matching::stored_match> matches =
+        ticker.engine.match_order(ticker.orderbook, order);
+    std::ranges::move(matches, std::back_inserter(accum_matches_));
+    return matches.size();
```
Comment (egelja) [inline exchange/src/exchange/tickers/manager/ticker_manager.cpp:51]:
```suggestion
    auto matches = ticker.engine.match_order(ticker.orderbook, order);
    size_t num_matches = matches.size();
    
    std::ranges::move(matches, std::back_inserter(accum_matches_));
    return num_matches;
```

can't access something after you move it.

**Review by egelja — CHANGES_REQUESTED** [review]

File: exchange/src/exchange/tickers/manager/ticker_manager.cpp (hunk 8 above)
Comment (stevenewald) [reply exchange/src/exchange/tickers/manager/ticker_manager.cpp:51]: good catch

File: exchange/src/exchange/tickers/engine/order_container.hpp (hunk 7 above)
Comment (stevenewald) [reply exchange/src/exchange/tickers/engine/order_container.hpp:35]: subjective, i prefer default

File: exchange/src/exchange/tickers/engine/level_update_generator.cpp (hunk 3 above)
Comment (stevenewald) [reply exchange/src/exchange/tickers/engine/level_update_generator.cpp:27]: good point, but this is checking to make sure we don't :)
This is asserted/handled elsewhere but this is just an extra sanity check, not really required

In more detail: std::round(price*100)/100 should be deterministic, i.e. 5.000001 and 5.00000002 should round to the exact same number

File: exchange/src/exchange/tickers/engine/order_container.hpp (hunk 7 above)
Comment (egelja) [reply exchange/src/exchange/tickers/engine/order_container.hpp:35]: i'm not being subjective here, default arguments tend to lead to issues with overloading

https://quuxplusone.github.io/blog/2020/04/18/default-function-arguments-are-the-devil/

Comment (gavinnwang) [reply exchange/src/exchange/tickers/engine/order_container.hpp:35]: @stevenewald  [https://stackoverflow.com/a/24171267](url) 
I prefer overloading in this case especially the default constructor for orderbook is used for testing purposes, it is good to separate them

Comment (stevenewald) [reply exchange/src/exchange/tickers/engine/order_container.hpp:35]: Ok sure let's go with overloading

**Comment by gavinnwang** [comment]
just resolved all comments related to code quality

**Comment by stevenewald** [comment]
good work @gavinnwang

**Pull request merged**